[Blog Post] Resources - Migrating Your React Native Library to the New Architecture (#3163)
facebook/react-native-website: website/blog/2022-06-16-resources-migrating-your-react-native-library-to-the-new-architecture.md

---

> title: Resources - Migrating Your React Native Library to the New Architecture

Reviewer: ```suggestion
title: Helping migrate React Native libraries to the New Architecture
```
this title is more aligned with the tone, also I'm not a big fan of the "Resources -" at the start, feels like it should almost be a tag instead

> author: "cipolleschi"

Reviewer 2: @cipolleschi You need to add yourself to `website/blog/authors.yml`

Author: thanks

tags: [announcement]
date: 2022-06-16
---

> **TL; DR**: We are working on improving the resources supporting the react Native New Architecture. We already released a repository to help migrating your app ([RNNewArchitectureApp](https://github.com/react-native-community/RNNewArchitectureApp)) and one for your libraries ([RNNewArchitectureLibraries](https://github.com/react-native-community/RNNewArchitectureLibraries)). We are also revamping the [New Architecture guide](https://github.com/facebook/react-native-website/pull/3037) on the Website and we created a [GitHub Working Group](https://github.com/reactwg/react-native-new-architecture/discussions) to answer questions related to the new Architecture.

Reviewer 2: ```suggestion
**tl;dr**: We are working on improving the resources supporting the React Native New Architecture. We already released a repository to help migrating your app ([RNNewArchitectureApp](https://github.com/react-native-community/RNNewArchitectureApp)) and one for your libraries ([RNNewArchitectureLibraries](https://github.com/react-native-community/RNNewArchitectureLibraries)). We are also revamping the [New Architecture guide](https://github.com/facebook/react-native-website/pull/3037) on the Website and we created a [GitHub Working Group](https://github.com/reactwg/react-native-new-architecture/discussions) to answer questions related to the New Architecture.
```

> **tl; dr**: We are working on improving the resources supporting the react Native New Architecture. We already released a repository to help migrating your app ([RNNewArchitectureApp](https://github.com/react-native-community/RNNewArchitectureApp)) and one for your libraries ([RNNewArchitectureLibraries](https://github.com/react-native-community/RNNewArchitectureLibraries)). We are also revamping the [New Architecture guide](https://github.com/facebook/react-native-website/pull/3037) on the Website and we created a [GitHub Working Group](https://github.com/reactwg/react-native-new-architecture/discussions) to answer questions related to the new Architecture.

Reviewer: ```suggestion
**tl; dr**: We are working on improving the resources supporting the React Native New Architecture. We already released a repository to help migrating your app ([RNNewArchitectureApp](https://github.com/react-native-community/RNNewArchitectureApp)) and one for your libraries ([RNNewArchitectureLibraries](https://github.com/react-native-community/RNNewArchitectureLibraries)). We are also revamping the [New Architecture guide](https://github.com/facebook/react-native-website/pull/3037) on the Website and we created a [GitHub Working Group](https://github.com/reactwg/react-native-new-architecture/discussions) to answer questions related to the new Architecture.
```

> 

Reviewer 3: ```suggestion

<!--truncate-->

```
We should add truncation marker somewhere, since the post is quite long. Not sure if it's an ideal spot, maybe we can leave also a bit of Intro.

## Introduction

In this post we are sharing an update on tools and resources to help React Native library developers to migrate their **Native Modules** and **Native Components** to their **New Architecture** equivalents, **TurboModule** and **Fabric Components**.

> React Native users leverage a vast number of open-source libraries for their apps. For a complete and consistent ecosystem, it is necessary that these libraries migrate such that everyone can benefit from the unlocked capabilities and performance improvements of the new architecture.

Author: Thanks! I swooped the document again looking for all the `new architecture`, `new Architecture` and `New architecture` combos and now it should be consistent! 😉

Here is what we’re working on to support library developers to migrate to the new architecture:

- **Documentation:** We are expanding the [New Architecture guide](https://github.com/facebook/react-native-website/pull/3037) on the website to cover more concepts of the new architectures and how to develop your components.
- **Example Migrations:** We’ve set up two repositories to demonstrate how to migrate a React Native app to the New Architecture ([RNNewArchitectureApp](https://github.com/react-native-community/RNNewArchitectureApp)) and how to create a **Fabric Component** and a **TurboModule** that work with both architectures ([RNNewArchitectureLibraries](https://github.com/react-native-community/RNNewArchitectureLibraries)).
- **Support:** Earlier this year, we created a [GitHub Working Group](https://github.com/reactwg/react-native-new-architecture/discussions) dedicated to discussion and questions around the New Architecture.

In this post, we will dig deeper into these resources and explain in more detail how you can use them most efficiently. Finally, we will provide a snapshot of the current migration state for the most used React Native libraries.

### Documentation

> In the past 6 months, we’ve added a guide on adopting the New Architecture (https://reactnative.dev/docs/new-architecture-intro) and an architecture deep-dive (https://reactnative.dev/architecture/overview) on Fabric. We plan to expand this to include more guides and documentation around creating TurboModules, understanding CodeGen, and more. We hope to have updates to share by the 0.70 release.

Reviewer 3: ```suggestion
In the past 6 months, we’ve added a guide on adopting the New Architecture (/docs/new-architecture-intro) and an architecture deep-dive (/architecture/overview) on Fabric. We plan to expand this to include more guides and documentation around creating TurboModules, understanding CodeGen, and more. We hope to have updates to share by the 0.70 release.
```

Let's use internal links for doc pages. Those prevents the whole app rerender on navigation which lead to more seamless and quicker page loads.

Reviewer 3: It also would be nice to add links to the certain words instead of listing the whole URL in content.
<img width="1047" alt="Screenshot 2022-06-17 at 09 24 04" src="https://user-images.githubusercontent.com/719641/174247433-15ce6eb8-3095-400f-a518-4deeb0f2d688.png">

Author: Thanks for the explanation about why it is useful to use internal links!
I missed these two spots for use words for links! 😱

> At the moment, the **New Architecture** guide covers how to migrate your app (https://reactnative.dev/docs/new-architecture-app-intro) and your libraries (https://reactnative.dev/docs/new-architecture-library-intro) to support the New Architecture properly.

Reviewer 3: ```suggestion
At the moment, the **New Architecture** guide covers how to migrate your app (/docs/new-architecture-app-intro) and your libraries (/docs/new-architecture-library-intro) to support the New Architecture properly.
```

Same as above.

Reviewer: ```suggestion
Currently, the **New Architecture** guide covers how to migrate your app (https://reactnative.dev/docs/new-architecture-app-intro) and your libraries (https://reactnative.dev/docs/new-architecture-library-intro) to support the New Architecture properly.
```
(nit: "at the moment" is an italian expression ;)

> At the moment, the **New Architecture** guide covers how to migrate your app (https://reactnative.dev/docs/new-architecture-app-intro) and your libraries (https://reactnative.dev/docs/new-architecture-library-intro) to support the New Architecture properly.

Reviewer: ```suggestion
Currently, the **New Architecture** guide covers how to migrate your app (https://reactnative.dev/docs/new-architecture-app-intro) and your libraries (https://reactnative.dev/docs/new-architecture-library-intro) to support the New Architecture properly.
```
nit: "at the moment" is mostly an italian expression ;)

Author: Aw, thanks! :D

If you are interested in the evolution of this guide, or have feedback, you can follow along on this pull request (https://github.com/facebook/react-native-website/pull/3037).

### Example Migrations

For developers who may want to follow along in code, we’ve prepared two example repositories.

> #### RNNewArchitectureApp

Reviewer 3: Generally adding links to the headers is something that we want to avoid, since they will always have the `#` link anyway and it might be confusing for the users.

> #### RNNewArchitectureApp

Author: Yup, added to the `This repo` fragment in the first sentence!

> This repo was created to demonstrate how to migrate an app, the native modules and the native components from the legacy architecture (0.67) to the most recent version React Native. Each commit corresponds to an isolated migration step.

Author: Amazing, thank you so much!

<figure>
    <img src="/blog/assets/new-arch-example-steps-to-migrate-an-app.png" alt="Example steps to migrate an app" />
    <figcaption>Commit list for a migration in the RNNewArchitectureApp repository</figcaption>
</figure>

> [Image: ]

Reviewer: ```suggestion
```
leftover?

The repo is organized as follows:

- A **main** branch has no code but a README.md which advertises other branches.
- Several migration branches which show a migration from a specific version of RN to another.

Some of the migration branches also have a **RUN.md** file which describes in a more human-readable fashion the exact steps that have been applied in every commit.

We plan to keep this example up to date with the most recent stable releases, adding migrations to any minor release of React Native we are going to release. If you notice issue with any of the steps, please file an issue in the repository. This will hold until we have the reasonable feeling that most of the React Native users have migrated to the New Architecture.

> #### RNNewArchitectureLibraries

Reviewer 4: As mentioned [here](https://github.com/facebook/react-native-website/pull/3163#discussion_r900062313), we don't want links in the headers.

The working group is organized in several sections:

- [Announcements](https://github.com/reactwg/react-native-new-architecture/discussions/categories/announcements): A place to share milestones and significant updates on the RN New Architecture Rollout
- [Deep Dive](https://github.com/reactwg/react-native-new-architecture/discussions/categories/deep-dive): A place to chat about deep dives and technical-specific topics
- [Documentation](https://github.com/reactwg/react-native-new-architecture/discussions/categories/documentation): A place to chat about the New Architecture documentation and migration material
- [Libraries](https://github.com/reactwg/react-native-new-architecture/discussions/categories/libraries): A place to chat about 3rd party libraries and their migration story to the New Architecture
- [Q&A](https://github.com/reactwg/react-native-new-architecture/discussions/categories/q-a): A place to ask the community for help on the New Architecture topics
- [Releases](https://github.com/reactwg/react-native-new-architecture/discussions/categories/releases): A place to chat about release specific bugs & build problems

To use this group effectively:

> - **Make sure your library is listed inside the [Libraries] (https://github.com/reactwg/react-native-new-architecture/discussions/categories/libraries)section**. This will help us share a status update on the migration of your library and will help us understand which struggles library authors are facing to support you better.

Reviewer: ```suggestion
- **Make sure your library is listed inside the [Libraries](https://github.com/reactwg/react-native-new-architecture/discussions/categories/libraries) section**. This will help us share a status update on the migration of your library and will help us understand which struggles library authors are facing to support you better.
```
fix

- **Leverage the Q&A [section](https://github.com/reactwg/react-native-new-architecture/discussions/categories/q-a) if you face a blocker and need support**. Our team and community experts are monitoring and will support at our best effort.
- **Keep an eye on the other sections for topics that may affect you**. A new release may introduce exactly the API that you were looking for. You can subscribe to particular discussions via Github.

We plan to support this group until the **New Architecture** is enabled by default and all the major libraries have been migrated to it.

> ### Migration Status of Certain Libraries

Reviewer: ```suggestion
### Migration Status of Popular Libraries
```

- [react-native-screens](https://github.com/reactwg/react-native-new-architecture/discussions/13): 🏃‍♂️ Ongoing
- [react-native-slider](https://github.com/reactwg/react-native-new-architecture/discussions/38): 🎬 Started
- [react-native-template-new-architecture](https://github.com/reactwg/react-native-new-architecture/discussions/21): ✅ Migrated. Gradually adopting/testing more companion Libraries
- [react-native-template-typescript](https://github.com/reactwg/react-native-new-architecture/discussions/22): ✅ Migrated
- [react-native-webview](https://github.com/reactwg/react-native-new-architecture/discussions/19): 🎬 Started

## Next Steps

We are invested in supporting the React Native community’s adoption of the new architecture. Concretely, we will continue to:

- Offer best-effort support in the **Working Group.**
- Provide more examples about how to achieve amazing results with the new architecture in the **RNNewArchitecture** repositories.
- Provide clear and up-to-date documentation on the **New Architecture**.
- Track the migration status of essential React Native libraries in the **Working Group**.
- Simplify the migration path for developers

> In addition, React Native 0.69 will ship with improved devX for app and library developers for new architecture adoption. You can find more information about the 0.69.0 release [here](https://github.com/reactwg/react-native-releases/discussions/21).

Reviewer: This phrase imply this blogpost will come out before 0.69.0 - @cortinico @fortmarek we should probably revise the blogpost draft to account for this blogpost existing (maybe?)

Personally I think we should avoid publishing two blogposts back-to-back, it makes it more likely that folks will miss out on one or the other; so if we want them to go out in the span of a few days, let's make sure that the other one has a link to the previous to raise awareness

Reviewer 4: Thanks for the callout @kelset - it'd be great to link to this from the 0.69 blogpost. If we could merge this by Monday and then ~ Wednesday release 0.69, that'd be a reasonable schedule imho. @cipolleschi wdyt? I don't want to rush this if you think this needs more time.

Author: Yes, sure! 

I'm apply all the comments and suggestions right now. On Monday I'll be on vacation (I had already planned to try and publish this ASAP!), but if everything is right, feel free to push the merge button!

---
title: Helping migrate React Native libraries to the New Architecture
authors: [cipolleschi]
tags: [announcement]
date: 2022-06-16
---

**tl; dr**: We are working on improving the resources supporting the React Native New Architecture. We already released a repository to help migrating your app ([RNNewArchitectureApp](https://github.com/react-native-community/RNNewArchitectureApp)) and one for your libraries ([RNNewArchitectureLibraries](https://github.com/react-native-community/RNNewArchitectureLibraries)). We are also revamping the [New Architecture guide](https://github.com/facebook/react-native-website/pull/3037) on the Website and we created a [GitHub Working Group](https://github.com/reactwg/react-native-new-architecture/discussions) to answer questions related to the New Architecture.

<!--truncate-->

## Introduction

In this post we are sharing an update on tools and resources to help you migrate your **Native Modules** and **Native Components** to their **New Architecture** equivalents, **TurboModule** and **Fabric Components**.

> React native users leverage vast number of open source libraries for building apps. For a complete and consistent ecosystem, it is necessary that these libraries migrate such that everyone can benefit from the unlocked capabilities and performance improvements of the New Architecture.

Reviewer 3: ```suggestion
React Native users leverage vast number of open source libraries for building apps. For a complete and consistent ecosystem, it is necessary that these libraries migrate such that everyone can benefit from the unlocked capabilities and performance improvements of the New Architecture.
```

Case fix, on which CI is currently failing on.
